sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+    // Initialize array using existing clang Initialization mechanisms
+    local.array = w_s; 
+    // Call accessor�s init function
+    Accessor::init(&local.outAcc, AccData, AccR1, AccR2, I);
+
+    callee(&local, id<1> wi);
+}
+```
+
+The sharp-eyed reviewer of `SemaSYCL.cpp` will notice that the array
+is actually double-wrapped in structs. This was done simply to preserve
+the interface to an existing function (`CreateAndAddPrmDsc`) which
+processes each kernel caller parameter as a capture object member. 
+The object being added to a list in `CreateAndAddPrmDsc` is `Fld`,
+which is expected to be a field of some struct. So a wrapped struct
+cannot be passed to this function. A double-wrapped struct is needed
+as shown below. This does not affect the generated code.
+
+```C++
+struct {
+  struct {
+    int array[100];
+  }
+}
+```
+
+This could be changed but it would mean changes to the `CreateAndAddPrmDsc`
+implementation, to all its callers and to the place where the list created
+by it is processed.
+By wrapping the array twice, the inner, single-wrapped array appears as a
+member of a struct and meets the requirements of the existing code.
+
+<h3>Fix 2: Kernel Arguments that are Arrays of Accessors</h3>
+
+Arrays of accessors are supported in a manner similar to that of a plain
+Accessor. For each accessor array element, the four values required to
+call its init function are passed as separate arguments to the kernel.
+Reassembly within the kernel caller is serialized by accessor array element.
+
+<h4>Source code fragment:</h4>
+
+```C++
+  myQueue.submit([&](handler &cgh) {
+    using Accessor =
+        accessor<int, 1, access::mode::read, access::target::global_buffer>;
+    Accessor inAcc[2] = {in_buffer1.get_access<access::mode::read>(cgh),
+                         in_buffer2.get_access<access::mode::read>(cgh)};
+    auto outAcc = out_buffer.get_access<access::mode::write>(cgh);
+
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = inAcc[0][index] + inAcc[1][index];
+    });
+  });
+```
+
+<h4>Integration header:</h4>
+
+```C++
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE20->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_accessor, 4062, 32 },
+  { kernel_param_kind_t::kind_accessor, 4062, 64 },
+};
+```
+
+<h4>Device code generated in pseudo-code form:</h4>
+
+```C++
+struct Capture {
+    Accessor outAcc; 
+    Accessor inAcc[2];
+    () {
+        // Body
+    }
+}
+
+spir_kernel void caller(
+    __global int* outAccData,    // args of OutAcc
+    range<1> outAccR1,
+    range<1> outAccR2,
+    id<1> outI, 
+    __global int* inAccData_0,   // args of inAcc[0]
+    range<1> inAccR1_0,
+    range<1> inAccR2_0,
+    id<1> inI_0,
+    __global int* inAccData_1,   // args of inAcc[1]
+    range<1> inAccR1_1,
+    range<1> inAccR2_1,
+    id<1> inI_1,
+)
+{
```

Comment:
BTW, what are the restrictions on how big the array or array of accessors can be? Maybe worth adding typical value here.

Reply:
At present there is no check on array size, just as there are no checks on the sizes of any other arguments such as structs.